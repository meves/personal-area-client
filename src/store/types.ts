export type ErrorMessage = { 
    message: string 
}

export type AuthToken = { 
    token: string 
}

export type UserFromList = {
    id: number
    firstname: string
    lastname: string
    createdAt: string
    updatedAt: string
}

export type DataWithUsers = {
    users: UserFromList[]
}

export type DataGreeting = {
    data: {
        greeting: {
            id: number
            greeting: string
        }
    }
}