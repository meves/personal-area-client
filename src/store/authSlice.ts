import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { AppDispatch, RootState } from "."
import { AuthRequests } from "../http/auth"
import { HTTP_CODES } from "../http/codes"
import jwt_decode from "jwt-decode";
import { ErrorMessage, AuthToken, AuthData } from "../types";


// initialState type
export interface AuthState {
    isAuth: boolean
    user: UserFromToken | null
    errorMessage: string
}
// initialState object
const initialState: AuthState = {
    isAuth: false,
    user: null,
    errorMessage: ""
}

// slise with reducers
export const authSlice = createSlice({
    name: 'auth',
    initialState,
    reducers: {
        setIsAuthAction: (state, action: PayloadAction<boolean>) => {
            state.isAuth = action.payload
        },
        setUserAction: (state, action: PayloadAction<UserFromToken | null>) => {
            state.user = action.payload
        },
        setAuthErrorMessageAction: (state, action: PayloadAction<string>) => {
            state.errorMessage = action.payload
        }
    }
})

// actions
export const { setIsAuthAction, setUserAction, setAuthErrorMessageAction } = authSlice.actions;

export default authSlice.reducer

// selectors
export const selectAuth = (state: RootState) => state.auth.isAuth
export const selectUser = (state: RootState) => state.auth.user
export const selectAuthErrorMessage = (state: RootState) => state.auth.errorMessage

// thunks
export class AuthThunk {

    static auth = () => 
        async (dispatch: AppDispatch) => {
            const token: string | null = localStorage.getItem("token");
            const data: AuthData = await AuthRequests.auth(token);
            if (data.code === "Success") {
                const user: UserFromToken = jwt_decode(data.token as string);
                await resultSigning(true, data.token, user, data.message, dispatch);
            } else {
                await resultSigning(false, data.token, null, data.message, dispatch);
            }
        }

    static signup = (email: string, password: string) => 
        async (dispatch: AppDispatch) => {
            const data: AuthData = await AuthRequests.signup(email, password);
            if (data.code === "Success") {
                const user: UserFromToken = jwt_decode(data.token);
                await resultSigning(true, data.token, user, data.message, dispatch);
            } else {
                await resultSigning(false, data.token, null, data.message, dispatch);
            }
        }

    static signin = (email: string, password: string) => 
        async (dispatch: AppDispatch) => {
            const data: AuthData = await AuthRequests.signin(email, password);
            if (data.code === "Success") {
                const user: UserFromToken = jwt_decode(data.token);
                await resultSigning(true, data.token, user, data.message, dispatch);
            } else {
                await resultSigning(false, data.token, null, data.message, dispatch);
            }
        }

    static logout = () => 
        async (dispatch: AppDispatch) => {
            localStorage.removeItem("token");
            dispatch(AuthThunk.auth());
        }    
}

// types
export type UserFromToken = {
    id: number
    email: string
    role: string
}

// utils
async function resultSigning(
    isAuth: boolean,
    token: string,
    user: UserFromToken | null,
    message: string,
    dispatch: AppDispatch
) {
    dispatch(setIsAuthAction(isAuth));
    dispatch(setUserAction(user));
    dispatch(setAuthErrorMessageAction(message));
    localStorage.setItem("token", token);
}
