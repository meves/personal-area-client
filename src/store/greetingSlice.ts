import { GreetingRequests } from './../http/greeting';
import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { AppDispatch, RootState } from ".";
import { HTTP_CODES } from '../http/codes';
import { DataGreeting } from './types';


export interface GreetingState {
    greeting: string | null
    error: Error | null
}

const initialState: GreetingState = {
    greeting: null,
    error: null
}

export const greetingSlice = createSlice({
    name: "greeting",
    initialState,
    reducers: {
        setGreeting: (state, action: PayloadAction<string | null>) => {
            state.greeting = action.payload
        },
        setError: (state, action: PayloadAction<Error | null>) => {
            state.error = action.payload
        }
    }
});

export const { setGreeting, setError } = greetingSlice.actions;

export default greetingSlice.reducer;

export const selectGreeting = (state: RootState) => state.greeting.greeting;
export const selectError = (state: RootState) => state.greeting.error;

// thunks
export class GreetingThunk {
    static getGreetingThunk = (id: number) => 
        async (dispatch: AppDispatch) => {
            const response = await GreetingRequests.getGreeting(id);
            if (response.status === HTTP_CODES.INTERNAL_SERVER_ERROR_500) {
                const { error } = await response.json();
                dispatch(setError(error));
                dispatch(setGreeting(null));
                return false;
            }
            if (response.status === HTTP_CODES.OK_200) {
                const { data }: DataGreeting = await response.json();
                dispatch(setGreeting(data.greeting.greeting));
                dispatch(setError(null));
                return true;
            }
        }
}