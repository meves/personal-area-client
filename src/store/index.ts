import { configureStore } from "@reduxjs/toolkit";
import authReducer from "./authSlice";
import usersReducer from "./usersSlice";
import greetingReducer from "./greetingSlice";

export const store = configureStore({
    reducer: {
        auth: authReducer,
        users: usersReducer,
        greeting: greetingReducer
    }
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
