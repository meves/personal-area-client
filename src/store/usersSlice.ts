import { UsersRequests } from './../http/users';
import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { AppDispatch, RootState } from "."
import { HTTP_CODES } from '../http/codes';
import { ErrorMessage, DataWithUsers, UserFromList } from './types';

export interface UsersState {
    users: UserFromList[]
    errorMessage: string
}
// initial state
const initialState: UsersState = {
    users: [],
    errorMessage: ""
}

// users slice
export const usersSlice = createSlice({
    name: "users",
    initialState,
    reducers: {
        setUsersAction: (state, action: PayloadAction<UserFromList[]>) => {
            state.users = action.payload;
        },
        addCreatedUserAction: (state, action: PayloadAction<UserFromList[]>) => {
            state.users = state.users.concat(action.payload);
        },
        deleteSelectedUserAction: (state, action: PayloadAction<number>) => {
            state.users = state.users.filter(user => user.id !== action.payload);
        },
        updateSeletedUserAction: (state, action: PayloadAction<UserFromList>) => {
            state.users = state.users.map(user => {
                if (user.id === action.payload.id) {
                    return action.payload
                }
                return user;
            });
        },
        setErrorMessageAction: (state, action: PayloadAction<string>) => {
            state.errorMessage = action.payload;
        }
    }
});

// actions
export const { setUsersAction, setErrorMessageAction, addCreatedUserAction, deleteSelectedUserAction, updateSeletedUserAction } 
= usersSlice.actions;

export default usersSlice.reducer;

// selectors
export const selectUsers = (state: RootState): UserFromList[] => state.users.users;

// thunks
export class UserListThunk {
    static getUsersThunk = () => 
        async (dispatch: AppDispatch) => {
            const response = await UsersRequests.getAllUsers();
            if (response.status === HTTP_CODES.INTERNAL_SERVER_ERROR_500) {
                const { message }: ErrorMessage = await response.json();
                dispatch(setErrorMessageAction(message));
                return;
            }
            if (response.status === HTTP_CODES.OK_200) {
                const { users }: DataWithUsers = await response.json();
                dispatch(setUsersAction(users));
                return;
            }
        }    
    static getUserThunk = (id: number) => 
        async (dispatch: AppDispatch) => {
            const response = await UsersRequests.getUser(id);
            if (response.status === HTTP_CODES.INTERNAL_SERVER_ERROR_500) {
                const { message }: ErrorMessage = await response.json();
                dispatch(setErrorMessageAction(message));
                return;
            }
            if (response.status === HTTP_CODES.OK_200) {
                const { users }: DataWithUsers = await response.json();
                dispatch(updateSeletedUserAction(users[0]));
                return;
            }
        }    
    static createUserThunk = (firstname: string, lastname: string) => 
        async (dispatch: AppDispatch) => {
            const response = await UsersRequests.createUser(firstname, lastname);
            if (response.status === HTTP_CODES.INTERNAL_SERVER_ERROR_500) {
                const { message }: ErrorMessage = await response.json();
                dispatch(setErrorMessageAction(message));
                return;
            }
            if (response.status === HTTP_CODES.CREATED_201) {
                const { users }: DataWithUsers = await response.json();
                dispatch(addCreatedUserAction(users));
                return;
            }
        }    
    static deleteUserThunk = (id: number) => 
        async (dispatch: AppDispatch) => {
            const response = await UsersRequests.deleteUser(id);
            if (response.status === HTTP_CODES.INTERNAL_SERVER_ERROR_500) {
                const { message }: ErrorMessage = await response.json();
                dispatch(setErrorMessageAction(message));
                return;
            }
            if (response.status === HTTP_CODES.NO_CONTENT_204) {
                dispatch(deleteSelectedUserAction(id));
                return;
            }
        }    
    static updateUserThunk = (id: number, firstname: string, lastname: string) => 
        async (dispatch: AppDispatch) => {
            const response = await UsersRequests.updateUser(id, firstname, lastname);
            if (response.status === HTTP_CODES.INTERNAL_SERVER_ERROR_500) {
                const { message }: ErrorMessage = await response.json();
                dispatch(setErrorMessageAction(message));
                return;
            }
            if (response.status === HTTP_CODES.NO_CONTENT_204) {
                dispatch(UserListThunk.getUserThunk(id));
                return;
            }
        }
}