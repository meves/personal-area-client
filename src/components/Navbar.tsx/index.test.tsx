import React from "react";
import ReactDOM from "react-dom/client";
import * as useHooks from "../../app/hooks";
import { Navbar } from ".";
import { UserFromToken } from "../../store/authSlice";
import { store } from "../../store";


let root: any = null;
let container: any = null;
beforeEach(() => {
    container = document.createElement("div");
    root = ReactDOM.createRoot(container);
});
afterEach(() => {
    root.unmount(container);
    container.remove();
    container = null; 
});

describe("Navbar component", () => {
    it("displays user email that was sent in token", () => {
        // define mock user returned from AuthSlice
        const mockUser: UserFromToken | null = { 
            id: 1, email: "user@mail.ru", role: "USER" 
        };
        // simulate selector to return mockUser
        const selectUser = jest.fn(() => (mockUser));
        // mock useAppSelector
        jest.spyOn(useHooks, "useAppSelector").mockImplementation(() => {
            return selectUser();
        });
        // mock useAppDispatch
        jest.spyOn(useHooks, "useAppDispatch").mockReturnValue(store.dispatch);
        // mock logout function
        const logout = jest.fn(() => {
            console.log(`User signed out`);
            
        });
        root.render(<Navbar/>);
        expect(container).toMatchSnapshot();
    });
});