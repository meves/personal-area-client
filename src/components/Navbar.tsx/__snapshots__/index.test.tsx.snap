// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component displays user email that was sent in token 1`] = `<div />`;
